File: estore-backend/src/main/java/com/estore/shopping/service/CartService.java
package com.estore.shopping.service;

import com.estore.catalog.entity.Product;
import com.estore.catalog.repository.ProductRepository;
import com.estore.exception.BadRequestException;
import com.estore.inventory.service.InventoryService;
import com.estore.exception.InsufficientStockException;
import com.estore.exception.ResourceNotFoundException;
import com.estore.shopping.dto.*;
import com.estore.shopping.entity.Cart;
import com.estore.shopping.entity.CartItem;
import com.estore.shopping.repository.CartItemRepository;
import com.estore.shopping.repository.CartRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CartService {

    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final InventoryService inventoryService;

    @Transactional(readOnly = true)
    public CartDTO getCartByUserId(Long userId) {
        Cart cart = cartRepository.findByUserIdWithItems(userId)
                .orElseGet(() -> createEmptyCart(userId));
        return toCartDTO(cart);
    }

    @Transactional
    public CartDTO addToCart(Long userId, AddToCartRequest request) {
        Cart cart = cartRepository.findByUserIdWithItems(userId)
                .orElseGet(() -> createEmptyCart(userId));

        Product product = productRepository.findByIdAndActiveTrue(request.getProductId())
                .orElseThrow(() -> new ResourceNotFoundException("Product", "id", request.getProductId()));

        // Check stock from Inventory (available = quantity - reserved)
        int available = inventoryService.getInventoryByProductId(product.getId()).getAvailableQuantity();
        if (available < request.getQuantity()) {
            throw new InsufficientStockException(product.getId(), request.getQuantity(), available);
        }

        Optional<CartItem> existingItem = cartItemRepository.findByCartIdAndProductId(cart.getId(), product.getId());

        if (existingItem.isPresent()) {
            // Update quantity
            CartItem item = existingItem.get();
            int newQuantity = item.getQuantity() + request.getQuantity();
            available = inventoryService.getInventoryByProductId(product.getId()).getAvailableQuantity();
            if (available < newQuantity) {
                throw new InsufficientStockException(product.getId(), newQuantity, available);
            }
            item.updateQuantity(newQuantity);
            cartItemRepository.save(item);
        } else {
            // Add new item
            CartItem newItem = CartItem.builder()
                    .cart(cart)
                    .product(product)
                    .quantity(request.getQuantity())
                    .unitPrice(product.getPrice())
                    .build();
            cart.addItem(newItem);
            cartRepository.save(cart);
        }

        return getCartByUserId(userId);
    }

    @Transactional
    public CartDTO updateCartItem(Long userId, UpdateCartItemRequest request) {
        Cart cart = cartRepository.findByUserIdWithItems(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Cart", "userId", userId));

        CartItem item = cartItemRepository.findById(request.getItemId())
                .orElseThrow(() -> new ResourceNotFoundException("CartItem", "id", request.getItemId()));

        if (!item.getCart().getId().equals(cart.getId())) {
            throw new BadRequestException("Cart item does not belong to this cart");
        }

        // Check stock
        Product product = item.getProduct();
        int available = inventoryService.getInventoryByProductId(product.getId()).getAvailableQuantity();
        if (available < request.getQuantity()) {
            throw new InsufficientStockException(product.getId(), request.getQuantity(), available);
        }

        item.updateQuantity(request.getQuantity());
        cartItemRepository.save(item);

        return getCartByUserId(userId);
    }

    @Transactional
    public void removeCartItem(Long userId, Long itemId) {
        Cart cart = cartRepository.findByUserIdWithItems(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Cart", "userId", userId));

        CartItem item = cartItemRepository.findById(itemId)
                .orElseThrow(() -> new ResourceNotFoundException("CartItem", "id", itemId));

        if (!item.getCart().getId().equals(cart.getId())) {
            throw new BadRequestException("Cart item does not belong to this cart");
        }

        cart.removeItem(item);
        cartRepository.save(cart);
    }

    @Transactional
    public void clearCart(Long userId) {
        Cart cart = cartRepository.findByUserIdWithItems(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Cart", "userId", userId));

        cartItemRepository.deleteAllByCartId(cart.getId());
    }

    private Cart createEmptyCart(Long userId) {
        // This should be handled by CustomerService during registration
        throw new ResourceNotFoundException("Cart", "userId", userId);
    }

    private CartDTO toCartDTO(Cart cart) {
        List<CartItemDTO> items = cart.getItems().stream()
                .map(this::toCartItemDTO)
                .collect(Collectors.toList());

        int totalItems = items.stream()
                .mapToInt(CartItemDTO::getQuantity)
                .sum();

        BigDecimal totalAmount = items.stream()
                .map(CartItemDTO::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return CartDTO.builder()
                .id(cart.getId())
                .userId(cart.getUser().getId())
                .items(items)
                .totalItems(totalItems)
                .totalAmount(totalAmount)
                .createdAt(cart.getCreatedAt())
                .updatedAt(cart.getUpdatedAt())
                .build();
    }

    private CartItemDTO toCartItemDTO(CartItem item) {
        return CartItemDTO.builder()
                .id(item.getId())
                .productId(item.getProduct().getId())
                .productName(item.getProduct().getName())
                .productImageUrl(item.getProduct().getImageUrl())
                .quantity(item.getQuantity())
                .availableStock(inventoryService.getInventoryByProductId(item.getProduct().getId()).getAvailableQuantity())
                .unitPrice(item.getUnitPrice())
                .subtotal(item.getSubtotal())
                .build();
    }
}

File: estore-frontend/src/core/services/cart.service.ts
import { API_ENDPOINTS, apiClient, toApiError, unwrapResponse } from './api';
import { Cart, CartItem, Product } from '@/shared/types';

interface BackendCartItemDTO {
  id: number;
  productId: number;
  productName: string;
  productImageUrl?: string;
  quantity: number;
  availableStock?: number;
  unitPrice: number;
  subtotal: number;
}

interface BackendCartDTO {
  id: number;
  userId: number;
  items: BackendCartItemDTO[];
  totalItems: number;
  totalAmount: number;
  createdAt?: string;
  updatedAt?: string;
}

const mapCartItem = (item: BackendCartItemDTO): CartItem => ({
  id: item.id,
  quantity: item.quantity,
  availableStock: item.availableStock,
  unitPrice: Number(item.unitPrice),
  product: {
    id: item.productId,
    name: item.productName,
    imageUrl: item.productImageUrl || '',
    price: Number(item.unitPrice),
    description: '',
    active: true,
    availableStock: 0,
  },
});

const mapCart = (cart: BackendCartDTO): Cart => ({
  id: cart.id,
  userId: cart.userId,
  items: cart.items?.map(mapCartItem) || [],
  createdAt: cart.createdAt,
});

export const CartService = {
  getCart: async (userId: number): Promise<Cart> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.CART(userId));
      const data = unwrapResponse<BackendCartDTO>(response);
      return mapCart(data);
    } catch (error) {
      throw toApiError(error, 'Failed to fetch cart');
    }
  },

  addToCart: async (userId: number, product: Product, quantity: number): Promise<Cart> => {
    try {
      const response = await apiClient.post(API_ENDPOINTS.CART_ADD, {
        userId,
        productId: product.id,
        quantity,
      });
      const data = unwrapResponse<BackendCartDTO>(response);
      return mapCart(data);
    } catch (error) {
      throw toApiError(error, 'Failed to add to cart');
    }
  },

  updateCartItem: async (itemId: number, quantity: number): Promise<Cart> => {
    try {
      const response = await apiClient.put(API_ENDPOINTS.CART_UPDATE, { itemId, quantity });
      const data = unwrapResponse<BackendCartDTO>(response);
      return mapCart(data);
    } catch (error) {
      throw toApiError(error, 'Failed to update cart');
    }
  },

  removeFromCart: async (itemId: number): Promise<void> => {
    try {
      await apiClient.delete(API_ENDPOINTS.CART_REMOVE(itemId));
    } catch (error) {
      throw toApiError(error, 'Failed to remove from cart');
    }
  },

  clearCart: async (userId: number): Promise<void> => {
    try {
      await apiClient.delete(API_ENDPOINTS.CART_CLEAR(userId));
    } catch (error) {
      throw toApiError(error, 'Failed to clear cart');
    }
  },
};
